pkg/auth/keyring.go

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2023. Nydus Developers. All rights reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package auth
+
+import (
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	defaultSessionName = "_ses.nydus"
+)
+
+type KeyRing struct {
+	lock sync.RWMutex
+}
+
+func NewKeyRing() *KeyRing {
+	return &KeyRing{}
+}
+
+func (k *KeyRing) Add(id, value string) (int, error) {
+	sessKeyID, err := unix.KeyctlJoinSessionKeyring(defaultSessionName)
+	if err != nil {
+		return 0, errors.Wrap(err, "unable to session key")
+	}
+
+	if err := k.modifyKeyringPerm(sessKeyID, 0xffffffff, 0x80000); err != nil {
+		return 0, errors.Wrap(err, "unable to mod keyring permissions")
+	}
+
+	k.lock.Lock()
+	defer k.lock.Unlock()
```

Comment:
Why do we need a lock here? Is `unix.AddKey` not atomic?

Reply:
We need this. `Addkey` only launch a `syscall`, but can't make sure it's atomic.
```shell
r0, _, e1 := Syscall6(SYS_ADD_KEY, uintptr(unsafe.Pointer(_p0)), uintptr(unsafe.Pointer(_p1)), uintptr(_p2), uintptr(len(payload)), uintptr(ringid), 0)
```